src/dataframe/dataframe.jl

```diff
@@ -1674,12 +1689,13 @@ function Base.push!(df::DataFrame, row::Union{AbstractDict, NamedTuple};
             S = typeof(val)
             T = eltype(col)
             if S <: T || !promote || promote_type(S, T) <: T
-                push!(col, val)
+                insert!(col, loc, val)
             else
                 newcol = similar(col, promote_type(S, T), targetrows)
-                copyto!(newcol, 1, col, 1, nrows)
-                newcol[end] = val
                 firstindex(newcol) != 1 && _onebased_check_error()
+                copyto!(newcol, 1, col, 1, loc-1)
+                copyto!(newcol, loc+1, col, loc, nrows-loc+1)
+                newcol[loc] = val
```

Comment:
In terms of cache locality, this would be more logical. Probably doesn't matter in practice though. (Same elsewhere.)
```suggestion
                copyto!(newcol, 1, col, 1, loc-1)
                newcol[loc] = val
                copyto!(newcol, loc+1, col, loc, nrows-loc+1)
```